Adding in a way of plugin defining route to non-MVC API Extensions

wilsonge opened the pull request:
Version 1 of getting a response. With this you now get a URL of http://localhost/~george/joomla-cms/api/index.php/article giving:

`{"links":{"self":"http:\/\/localhost\/~george\/joomla-cms\/api\/index.php\/article"}}`

Next step is to have the plugin inject a mapping object which runs though and maps the object to the response

wilsonge commented on api/components/com_content/dispatcher.php:
```diff
@@ -0,0 +1,97 @@
+		$namespace = rtrim($this->namespace, '\\') . '\\';
+
+		// Set up the client
+		$client = 'Site';
```
Only difference here between this and the site dispatcher is forcing the client to site for the controller creation

wilsonge commented on libraries/src/CMS/Application/ApiApplication.php:
```diff
@@ -267,4 +266,19 @@ public function dispatch($component = null)
+	 */
+	public function getMenu($name = null, $options = array())
+	{
+		return null;
```
Not sure what to do about this. This was a hack for working around a menu item being required at: https://github.com/joomla-projects/gsoc17_webservices/blob/4.0-dev/libraries/src/CMS/Controller/Controller.php#L750

wilsonge replied:
Pathway also relies on the menu existing. Unsure whether to make that return `null` too

wilsonge commented on libraries/src/CMS/Controller/Controller.php:
```diff
@@ -852,6 +852,12 @@ public function getView($name = '', $type = '', $prefix = '', $config = array())
 			{
 				$prefix = $this->app->getName();
+
+				if ($prefix === 'api')
```
Hack around to use the right model - this should be moved to the MvcFactory

wilsonge commented on libraries/src/CMS/Document/Factory.php:
```diff
@@ -32,6 +32,11 @@ public function createDocument($type = 'html', array $attributes = array())
 		$type  = preg_replace('/[^A-Z0-9_\.-]/i', '', $type);
 		$ntype = null;
 
+		if ($type === 'applicationvnd.apijson')
```
Ensure's the right JDocument class is loaded (from the JInput format)

wilsonge commented on libraries/joomla/document/api/jsonapi.php:
```diff
@@ -26,7 +26,7 @@ class JDocumentApiJsonapi extends JDocumentJson implements JsonSerializable
 	 * @since  __DEPLOY_VERSION__
 	 */
-	protected $document;
+	public $document;
```
Do we want to keep this protected and just have public wrapping methods for everything? Or do we want to make it public?

wilsonge commented on libraries/joomla/document/jsonapi.php:
```diff
@@ -26,7 +26,7 @@ class JDocumentJsonapi extends JDocumentJson implements JsonSerializable
 	 * @since  __DEPLOY_VERSION__
 	 */
-	protected $document;
+	public $document;
```
We need to decide whether to create wrapper methods for everything (like the `addLink` method I'm using in com_content view, or make this public and delete any proxies to that method)

fastslack replied:
I think there is no problem to leave document public

wilsonge replied:
I think it depends how much you want to rely on the implementation. For example we can add wrapper interfaces on some of the library interfaces and swap it out if we ever need to. If you make this public we're totally dependent on the library for life

wilsonge commented on libraries/joomla/document/jsonapi.php:
```diff
@@ -41,6 +41,7 @@ public function __construct($options = array())
 
 		// Set mime type to JSON-API
 		$this->_mime = 'application/vnd.api+json';
+		$this->_type = 'jsonapi';
```
We need to decide on whether to make this type explicit or not as it defines what the View Format is (i.e. Should component's be limited to a single JSON view or should be have json+ld and jsonapi and json views all available?

fastslack replied:
For first instancese we can use fixed type but for future would be awesome to have some switch for differents json types

wilsonge commented on api/components/com_content/dispatcher.php:
```diff
@@ -0,0 +1,67 @@
+	{
+		if ($client === null || strtolower($client) === 'api')
+		{
+			$client = 'Site';
```
Hack to force using the site controller for com_content to test things working

wilsonge commented:
It looks like you're trying to go patching patch. I've fixed some of the stuff here into the main lib_api branch as I fixed it. Like there's no changes to the libraries/src/CMS/Controller/Controller.php file here anymore :/

wilsonge commented on libraries/src/CMS/Mvc/Factory/MvcFactory.php:
```diff
@@ -62,6 +62,11 @@ public function __construct($namespace, CMSApplicationInterface $application)
 	 */
 	public function createModel($name, $prefix = '', array $config = array())
 	{
+		if (empty($prefix) && $this->application->isClient('api'))
```
Dirty hack to ensure that models are correctly loaded. We should use our own ApiFactory
